src/charm.py

```diff
@@ -32,140 +31,144 @@
-        self.framework.observe(self.on.receive_remote_write_relation_changed, self._configure)
-        self.framework.observe(self.on.receive_remote_write_relation_broken, self._configure)
-        self.framework.observe(self.metrics_consumer.on.targets_changed, self._configure)
-        self.framework.observe(self.alertmanager_consumer.on.cluster_changed, self._configure)
-        self.framework.observe(
-            self.on.validate_configuration_action, self._on_validate_configuration
+        self.alertmanager_consumer = AlertmanagerConsumer(
+            charm=self,
+            relation_name="alertmanager",
+        )
+
+        self._register_event_handlers()
+
+    @property
+    def _scrape_job(self):
+        return [{"static_configs": [{"targets": [f"*:{self._port}"]}]}]
+
+    @property
+    def _prometheus_layer(self) -> Layer:
+        """Construct the pebble layer.
+
+        Returns:
+            a Pebble layer specification for the Prometheus workload container.
+        """
+        logger.debug("Building pebble layer")
+        layer_config = {
+            "summary": "Prometheus layer",
+            "description": "Pebble layer configuration for Prometheus",
+            "services": {
+                self._name: {
+                    "override": "replace",
+                    "summary": "prometheus daemon",
+                    "command": self._command(),
+                    "startup": "enabled",
+                }
+            },
+        }
+
+        return Layer(layer_config)
+
+    @property
+    def _external_url(self) -> str:
+        """Return the external hostname to be passed to ingress via the relation.
+
+        If we do not have an ingress, then use the pod ip as hostname.
+        The reason to prefer this over the pod name (which is the actual
+        hostname visible from the pod) or a K8s service, is that those
+        are routable virtually exclusively inside the cluster (as they rely)
+        on the cluster's DNS service, while the ip address is _sometimes_
+        routable from the outside, e.g., when deploying on MicroK8s on Linux.
+        """
+        if web_external_url := self.model.config.get("web_external_url"):
+            return web_external_url
+        if ingress_url := self.ingress.url:
+            return ingress_url
+        return f"http://{socket.getfqdn()}:{self._port}"
+
+    def _patch_kubernetes(self):
+        self.service_patch = KubernetesServicePatch(self, [(f"{self.app.name}", self._port)])
+        self.resources_patch = KubernetesComputeResourcesPatch(
+            self,
+            self._name,
+            resource_reqs_func=lambda: self._resource_reqs_from_config(),
         )
-        self.framework.observe(self.on.update_status, self._update_status)
-        self.framework.observe(self.resources_patch.on.patch_failed, self._on_k8s_patch_failed)
+
+    def _register_event_handlers(self):
```

Comment:
- This makes sense to me for lifecycle (charm) events.
- Style wise I personally prefer observing _custom_ events immediately next to the charm lib they are coming from.

Reply:
I'm a bit skeptic to the current length of the init class, but I'll revert it for now. The dict of arrays structure suggested by @rbarry82 a bit further down is a no-go from my PoV as it both introduces nested for loops and makes the code more verbose. Especially as the goal is to not have common entry points for events, which would mean every subscription would be three lines long.